File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/aws/ssm/Parameters.java
package uk.nhs.nhsx.core.aws.ssm;

import java.util.Set;
import java.util.function.Function;

public interface Parameters {

    Set<String> positive = Set.of("yes", "true", "enabled");

    <T> Parameter<T> parameter(ParameterName name, Function<String, T> convert);

    default <T extends Enum<T>> Parameter<T> ofEnum(ParameterName name, Class<T> type, T whenError) {
        return parameter(name, (v) -> convertEnum(type, v, whenError));
    }

    default <T extends Enum<T>> T convertEnum(Class<T> type, String value, T whenError) {
        try {
            return Enum.valueOf(type, value.toUpperCase());
        } catch (IllegalArgumentException noMatchingEnumValue) {
            return whenError;
        }
    }

    default Parameter<Boolean> ofBoolean(ParameterName name) {
        return parameter(name, (v) -> v!= null && positive.contains(v.toLowerCase()));
    }

    default Parameter<String> ofString(ParameterName name) {
        return parameter(name, Function.identity());
    }
}

File: src/aws/lambdas/incremental_distribution/cta/src/main/java/uk/nhs/nhsx/core/aws/ssm/Parameters.kt
package uk.nhs.nhsx.core.aws.ssm

import java.lang.Enum.valueOf
import java.util.*
import java.util.function.Function

interface Parameters {
    fun <T> parameter(name: ParameterName, convert: Function<String, T>): Parameter<T>

    companion object {
        val positive = setOf("yes", "true", "enabled")
    }
}

fun <T : Enum<T>> Parameters.ofEnum(name: ParameterName, type: Class<T>, whenError: T): Parameter<T> =
    parameter(name) { v: String -> convertEnum(type, v, whenError) }

fun <T : Enum<T>> convertEnum(type: Class<T>, value: String, whenError: T): T = try {
    valueOf(type, value.uppercase(Locale.getDefault()))
} catch (noMatchingEnumValue: IllegalArgumentException) {
    whenError
}

fun Parameters.ofBoolean(name: ParameterName): Parameter<Boolean> =
    parameter(name) { v: String? -> v != null && Parameters.positive.contains(v.lowercase(Locale.getDefault())) }
